SwiftCompilerSources/Sources/Optimizer/ModulePasses/MandatoryPerformanceOptimizations.swift

```diff
@@ -48,28 +50,32 @@ private func optimizeFunctionsTopDown(using worklist: inout FunctionWorklist,
 }
 
 private func optimize(function: Function, _ context: FunctionPassContext) {
-  runSimplification(on: function, context, preserveDebugInfo: true) { instruction, simplifyCtxt in
-    if let i = instruction as? OnoneSimplifyable {
-      i.simplify(simplifyCtxt)
-      if instruction.isDeleted {
-        return
+  var changed = true
+  while changed {
+    changed = false
+    
+    changed = changed || runSimplification(on: function, context, preserveDebugInfo: true) { instruction, simplifyCtxt in
+      if let i = instruction as? OnoneSimplifyable {
+        i.simplify(simplifyCtxt)
+        if instruction.isDeleted {
+          return
+        }
+      }
+      switch instruction {
+      case let apply as FullApplySite:
+        inlineAndDevirtualize(apply: apply, context, simplifyCtxt)
+      default:
+        break
       }
     }
-    switch instruction {
-    case let apply as FullApplySite:
-      inlineAndDevirtualize(apply: apply, context, simplifyCtxt)
-    default:
-      break
-    }
-  }
-
-  _ = context.specializeApplies(in: function, isMandatory: true)
-
-  removeUnusedMetatypeInstructions(in: function, context)
-
-  // If this is a just specialized function, try to optimize copy_addr, etc.
-  if context.optimizeMemoryAccesses(in: function) {
-    _ = context.eliminateDeadAllocations(in: function)
+    
+    changed = changed || context.specializeApplies(in: function, isMandatory: true)
```

Comment:
The `changed = changed || ...` does not work, because if `changed` is already true then the right hand side is not evaluated

Reply:
Ah, of course...